src/Symfony/Component/Validator/Constraints/IsinValidator.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+/**
+ * @author Laurent Masforné <l.masforne@gmail.com>
+ *
+ * @see https://en.wikipedia.org/wiki/International_Securities_Identification_Number
+ */
+class IsinValidator extends ConstraintValidator
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Isin) {
+            throw new UnexpectedTypeException($constraint, Isin::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = strtoupper($value);
+        $canonicalize = str_replace(' ', '', $value);
+
+        if (!(Isin::VALIDATION_LENGTH === \strlen($canonicalize))) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Isin::INVALID_LENGTH_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (!preg_match(Isin::VALIDATION_PATTERN, $value)) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Isin::INVALID_PATTERN_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (!$this->isCorrectChecksum($value)) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Isin::INVALID_CHECKSUM_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        return $value;
+    }
+
+    private function isCorrectChecksum($input)
+    {
+        $characters = str_split($input);
+        foreach ($characters as $i => $char) {
+            $characters[$i] = \intval($char, 36);
+        }
+        $checkDigit = array_pop($characters);
+        $number = implode('', $characters);
+        $expectedCheckDigit = $this->getCheckDigit($number);
+
+        return $checkDigit === $expectedCheckDigit;
+    }
+
+    /**
+     * This method performs the luhn algorithm
+     * to obtain a check digit.
+     */
+    private function getCheckDigit($input)
+    {
+        $input = (string) $input;
```

Comment:
For me, this line is unnecessary because it's a private function and the `$number` variable is always a sting (see [this](https://www.php.net/manual/en/function.implode.php)).

Reply:
Yes it's right, i have fixed this